### DISCUSSION TITLE
Dexie Typescript Models

### DISCUSSION THREAD

**Discussion by noelheesen** [dexie/Dexie.js#1578] [Q&A] [answered]
I wish to add new items without an id and retrieve them with an id.

I have the following code which auto increments id
```
interface Item {
  id: number,
  text: string
}

class DB extends Dexie {
  items!: Dexie.Table<Item, number>
  
  constructor() {
    this.version(1).stores({
      items: "++id",
    })
  }
}
```

When I want to add a new item I have to specify an ID which I don't have and will auto increment anyways, is there a way I can make this happen?

We know that dexie will auto generate an ID so why ask for it?
![afbeelding](https://user-images.githubusercontent.com/18031130/172019559-657bbb81-8108-47b3-a36b-b31988abc683.png)

Why do I need to check for an ID when I just retrieved the item by id?
![afbeelding](https://user-images.githubusercontent.com/18031130/172019718-28d958c6-39d5-4727-a851-d350e5b637e0.png)

Does anyone know a solution to this?

**Comment by dfahlander** [comment, marked as answer]
In Dexie 4.0 alpha (`npm install dexie@next`) you can achieve exactly that. Declare table as Table<Item, "id"> instead.

**Reply by noelheesen** [reply]
Thanks I'll go do this right away!

**Reply by noelheesen** [reply]
I think something broke in the alpha, I still get the same error and by the looks of it only method props are omitted?

![afbeelding](https://user-images.githubusercontent.com/18031130/172023641-a225ded5-fc9f-44fd-99eb-bd5ce9548a03.png)

**Reply by Kirbinator001** [reply]
I'm also trying to do this but running into a similar  issue. I've included what I've got below and the error message I'm getting.
```typescript
import { Dexie, Entity, Table } from 'dexie';

export class Item extends Entity<ItemsDB> {
  id!: number;
  text!: string;
}

export class ItemsDB extends Dexie {
  items!: Table<Item, 'id'>;

  constructor() {
    super('ItemsDB');
    this.version(1).stores({
      items: '++id',
    });

    this.items.mapToClass(Item);
  }
}

export const db = new ItemsDB();

db.on('populate', populate);

export async function populate() {
  await db.items.add({ text: 'Test' });
}
```

![image](https://user-images.githubusercontent.com/48194118/179458965-2ba34867-1476-4d17-9e39-37c2a02858d1.png)

**Reply by dfahlander** [reply]
Thanks for sharing. The plan is to use template literal types to detect the "++" in the declaration. For now, you can still achieve it by supplying the optional input properties as the third template argument:

```ts
export class ItemsDB extends Dexie {
  items!: Table<Item, 'id', 'id'>;
}
```
Sorry for missing to point that out in my previous post!
